Abstract common j-job tasks

WalterKolczynski-NOAA opened the pull request:
**Description**
Takes all of the tasks that are common to all j-jobs and abstracts them out into a shared script that is sourced by each job:
- Set and create $DATA directory
- Call setpy and set $cycle
- Set pid, pgmout, and pgmerr
- Source config files
- Source machine environment file

The common j-job header script is called by passing the job name for the `${machine}.env` files using the `-e` option, and a list of config files to source with the `-c` option.
```
${HOMEgfs}/ush/jjob_header.sh -e [env_job] -c "[config1 [config2 [...]]]"
```
The job name argument (`-e`) is mandatory, and the config list is optional but recommend to always use as well.

Some pre j-job rocoto entry scripts (`jobs/rocoto/*`) are currently doing much more than they should be. These sometimes required extra finagling, usually pre-calling the jjob header in the rocoto script before it does something.

Refs: #1069 

**Type of change**
- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**
- [x] Cycled test on Orion
- [x] Cycled test on Hera
- [x] Cycled test on WCOSS2
- [x] Coupled test on Orion
- [x] Coupled test on Hera
- [x] Couple test on WCOSS2

Post sounding (bufr) and gempak jobs complete successfully
AWIPS and WAFS failing for unrelated reasons
GDAS App jobs not tested
Some j-jobs are not called in the development workflow. TODOs an corresponding issues have been added to these.

**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

**Dependencies**
#1212
#1214
#1216
#1218

WalterKolczynski-NOAA committed:
Abstract common j-job tasks

Takes all of the tasks that are common to all j-jobs and abstracts
them out into a shared script that is sourced by each job:
- Set and create $DATA
- Call setpy and set $cycle
- Set pid, pgmout, and pgmerr
- Source config files
- Source machine environment file

The common j-job header script is called by passing a list of config
files to source:
```
${HOMEgfs}/ush/jjob_header.sh [config1 [config2 [...]]]
```

Some pre j-job rocoto entry scripts (`jobs/rocoto/*`) are currently
doing much more than they should be. These sometimes required extra
finagling, usually pre-calling the jjob header in the rocoto script
before it does something.

WalterKolczynski-NOAA committed:
Clean $DATA and have all steps of ocean anl use same $DATA

The ocean analysis jobs appear to need a persistent $DATA directory,
so the j-jobs have had their previous settings for $DATA restored.

Additionally, the j-job header now wipes any existing $DATA
directory if the variable $WIPE_DATA is set to "YES", which is the
default. To allow the persistent $DATA for the ocean analysis jobs,
the RUN and POST j-jobs set $WIPE_DATA to "NO".

WalterKolczynski-NOAA committed:
Only add CPU ranks to gempak CFP file when needed

The gempak j-job was always adding the CPU ranks to the CFP command
file. These are only needed (or allowed) on slurm, which prevented
the job from running properly on WCOSS2. Following the paradigm of
existing jobs, now the ranks are only added if `$CFP_MP` is "YES".

WalterKolczynski-NOAA committed:
Fix linter complaint in special gempak

Fix one minor thing to shut the linter up even though the line was
not touched in this PR.

WalterKolczynski-NOAA commented:
First wave of testing passes on all machines. Pulled in develop changes and running a last test on Orion where I'm going to compare output against develop.

WalterKolczynski-NOAA committed:
Add CFP_MP changes to env for gempak

These were meant to be part of 0d45879

WalterKolczynski-NOAA commented:
Output confirmed identical to develop. Had to revert PR #1158 locally on both to get the analysis to run successfully, at least with the chosen date.

aerorahul commented on jobs/JGDAS_ATMOS_ANALYSIS_DIAG:
```diff
@@ -1,49 +1,7 @@
-##########################################
-# Source machine runtime environment
-##########################################
-. ${HOMEgfs}/env/${machine}.env anal
```
This job is passing `anal` as the argument for sourcing `$machine.env` file.  In the `jjob_header.sh`, it is going to use `analdiag` as that is set in the rocoto job `analdiag.sh`. There is no entry for `analdiag` in `$machine.env`, atleast not in HERA.env

WalterKolczynski-NOAA replied:
Not in `ORION.env` either. Must not be important, since it completes and produces bit-identical results.

WalterKolczynski-NOAA replied:
(That's not a reason not to add an entry for analdiag, just maybe it should be empty?)

WalterKolczynski-NOAA replied:
Refactored the jjob header to take in the job name to pass to the environment script.

aerorahul commented on jobs/JGDAS_ATMOS_GLDAS:
```diff
@@ -1,42 +1,7 @@
-    [[ ${status} -ne 0 ]] && exit "${status}"
-done
-
+source "${HOMEgfs}/ush/jjob_header.sh" base gldas
```
This is going to create the `DATA` directory.
If `gldas` is going to be skipped, we should remove the empty `DATA` directory as well.

WalterKolczynski-NOAA replied:
Moved to after the cycle checks.

WalterKolczynski-NOAA replied:
Had to move it back because the script needs variables set in the gldas config files to determine whether to skip. Instead added removal of `$DATA` if aborting.

aerorahul commented on jobs/JGDAS_ATMOS_VERFOZN:
```diff
@@ -4,36 +4,12 @@
-export pid=$$
-export pgmout="OUTPUT.${pid}"
-export pgmerr=errfile
+source "${HOMEgfs}/ush/jjob_header.sh" base vrfy
```
This job does not source any configs.  We are now forcing it to source these files that may or may not be applicable for this job.
Likewise, it is going to source vrfy section in $machine.env

WalterKolczynski-NOAA replied:
It did, at least in development, but it was being done in the rocoto pre-job (`jobs/rocoto/vrfy.sh`) (and still is, for now). Added it here to be consistent with other j-jobs, since the job should be broken up so that only one j-job is called per rocoto task.

aerorahul commented on jobs/JGDAS_ENKF_DIAG:
```diff
@@ -1,49 +1,7 @@
-. ${HOMEgfs}/env/${machine}.env eobs
-status=$?
-[[ ${status} -ne 0 ]] && exit ${status}
+source "${HOMEgfs}/ush/jjob_header.sh" base anal eobs analdiag ediag
```
LHS uses `eobs` in $machine.env. 
RHS will be using ediag.

WalterKolczynski-NOAA replied:
Refactored the jjob header to take in the job name to pass to the environment script.

aerorahul commented on jobs/JGDAS_GLOBAL_OCEAN_ANALYSIS_POST:
```diff
@@ -1,41 +1,18 @@
 export STRICT="NO"
 source "${HOMEgfs}/ush/preamble.sh"
+WIPE_DATA="NO"
+DATA="${DATA:-${DATAROOT}/${RUN}ocnanal_${cyc}}"
```
why can't this be:
DATA="${DATAROOT}/${RUN}ocnanal_${cyc}}"
There is no one upstream setting `DATA`

WalterKolczynski-NOAA replied:
It can be. I just copied what was there before.

aerorahul reviewed (changes requested):
In spirit, this is a neat feature.
However, there are quite a few realizations where standardization likely hurts in the transparency of the jobs.

WalterKolczynski-NOAA committed:
Specify job for environment script

The j-job header is altered to require the specification of the job
name to pass to the environment script rather than using the inherent
job name since the ECFlow names are different and some jjobs call the
environment script with different names.

J-job header arguments are now passed in with options, with `-e`
setting the job name to provide the environment script (required) and
`-c` passing the list of config files (the former positional arguments).
A few jobs not yet in the dev workflow were not previously calling the
environment script. For those, I made up new job names as placeholders
(the environment scripts accept unrecongnized names).

Also added error messages for fatal errors and expanded documentation
a bit more.

WalterKolczynski-NOAA committed:
Remove ability to override $DATA for ocean analysis

The ocean analysis jobs were accepting previous values of `$DATA` if
it were set, which is not necessary.

aerorahul commented on ush/jjob_header.sh:
```diff
@@ -0,0 +1,114 @@
+shift $((OPTIND-1))
+
+if [[ -z ${env_job} ]]; then
+    echo "FATAL [${BASH_SOURCE[0]}]: Must specify a job name with -e"
```
I think an `exit 1` is missing after the echo.

WalterKolczynski-NOAA replied:
Fixed.

WalterKolczynski-NOAA committed:
Fix reading of config array in jjob header

WalterKolczynski-NOAA committed:
Fix reading in array of config files

WalterKolczynski-NOAA committed:
Fix GLDAS cycle checking

Previous commit moving the jjob_header after checking the cycle didn't
work because variables from the config files are needed. Instead, moved
the header back where it was and delete `$DATA` if we are aborting due
to the cycle.

WalterKolczynski-NOAA committed:
Add missing exit when job is not provided to header

There was no non-zero exit in the jjob header after printing the fatal
error message when the job name for the environment file is not provided
with `-e`.

WalterKolczynski-NOAA commented:
Testing complete after addressing reviewer comments. Output still identical.

aerorahul commented on jobs/JGFS_ATMOS_AWIPS_20KM_1P0DEG:
```diff
@@ -1,30 +1,11 @@
-# Set up the cycle variable
-######################################
-export cycle="t${cyc}z"
+export OMP_NUM_THREADS=${OMP_NUM_THREADS:-1}
```
shouldn't this be in the ex-script?

WalterKolczynski-NOAA replied:
Maybe, but this was here before (even though git diff is marking it as an addition; it was L5 before).

aerorahul commented on jobs/JGFS_ATMOS_AWIPS_G2:
```diff
@@ -1,35 +1,17 @@
-###########################################
-setpdy.sh
-. ./PDY
+export OMP_NUM_THREADS=${OMP_NUM_THREADS:-1}
```
same here.

WalterKolczynski-NOAA replied:
So was this. There is still much j-job cleaning to do.

aerorahul replied:
indeed.

aerorahul commented on jobs/rocoto/awips.sh:
```diff
@@ -94,14 +83,14 @@ for fhr in $fhrlst; do
 	    export fcsthrs=$fhr3
-	    export job="jgfs_awips_f${fcsthrs}_20km_${cyc}"
-	    export DATA="${DATAROOT}/$job"
+	    # export job="jgfs_awips_f${fcsthrs}_20km_${cyc}"
+	    # export DATA="${DATAROOT}/$job"
```
are we sure this is not going to bomb?

WalterKolczynski-NOAA replied:
No, but the AWIPS jobs aren't working right now anyway for unrelated reasons. It isn't causing any immediate issues up until that point. But that is why I left them as comments for now instead of just deleting them outright, so we can see how it was supposed to work before when we get a chance to fix the AWIPS job.

aerorahul replied:
understood.  thanks for explaining.

aerorahul commented on jobs/rocoto/awips.sh:
```diff
@@ -94,14 +83,14 @@ for fhr in $fhrlst; do
 	if [[ $(expr $fhr % 6) -eq 0 ]]; then
-	    export job="jgfs_awips_f${fcsthrs}_${cyc}"
-	    export DATA="${DATAROOT}/$job"
+	    # export job="jgfs_awips_f${fcsthrs}_${cyc}"
```
same here

aerorahul commented on jobs/rocoto/awips.sh:
```diff
@@ -131,12 +120,12 @@ for fhr in $fhrlst; do
 	    export fcsthrs=$fhr3
-	    export job="jgfs_awips_f${fcsthrs}_20km_${cyc}"
-	    export DATA="${DATAROOT}/$job"
+	    # export job="jgfs_awips_f${fcsthrs}_20km_${cyc}"
```
and here.

aerorahul reviewed (approved):
Changes look great.
A lot of duplicate lines have been eliminated.
I like that the header takes in named arguments that improves readability.

A few inline comments could use some explaining.
